README.md
# SixArm.com → Ruby → <br> Ramp is a toolkit of Ruby extensions

<!--header-open-->

[![Gem Version](https://badge.fury.io/rb/sixarm_ruby_ramp.svg)](http://badge.fury.io/rb/sixarm_ruby_ramp)
[![Build Status](https://travis-ci.org/SixArm/sixarm_ruby_ramp.png)](https://travis-ci.org/SixArm/sixarm_ruby_ramp)
[![Code Climate](https://api.codeclimate.com/v1/badges/750d63199bda82888002/maintainability)](https://codeclimate.com/github/SixArm/sixarm_ruby_ramp/maintainability)

* Git: <https://github.com/SixArm/sixarm_ruby_ramp>
* Doc: <http://sixarm.com/sixarm_ruby_ramp/doc>
* Gem: <https://rubygems.org/gems/sixarm_ruby_ramp>
* Contact: Joel Parker Henderson, <joel@sixarm.com>
* Project: [changes](CHANGES.md), [license](LICENSE.md), [contributing](CONTRIBUTING.md).

<!--header-shut-->


## Introduction

Ramp is a library of extensions to Ruby base classes.

We extend Array, Date, Enumerable, Hash, Kernel, Numeric, Object, Process, String, Time, and YAML.

For docs go to <http://sixarm.com/sixarm_ruby_ramp/doc>

Want to help? We're happy to get pull requests.


<!--install-open-->

## Install

### Gem

To install this gem in your shell or terminal:

    gem install sixarm_ruby_ramp

### Gemfile

To add this gem to your Gemfile:

    gem 'sixarm_ruby_ramp'

### Require

To require the gem in your code:

    require 'sixarm_ruby_ramp'

<!--install-shut-->


## Array

* car, cdr: aka first, rest (see shifted)
* choice, choices: one or more random elements from an array
* cross: return the cross pairings of an array with another array
* divvy: divides an array, like a pie, into a specified number of slices (deprecated)
* onto: return a hash that maps an array's keys on to another array's values
* rotate: moves the first element of an array to the end
* rest: return the rest of the items of the array (aka cdr, aka shifted)
* shifted, shifted!: return an array with the first n items shifted (aka cdr, aka rest)
* shuffle, shuffle!: randomly sort an array efficiently (backport)
* slices: divide an array into specified number of equal size sub-arrays (deprecated)
* to_csv: join a 2D array to a string as CSV (Comma Separated Values)
* to_tdf: join a 2D array to a string as TDF (Tab Delimited Format)
* to_tsv: join a 2D array to a string as TSV (Tab Separated Values)
* union: set union of the sub-arrays.


### Array join

* join: enhances #join with optional prefix, suffix, and infix.
* join_prefix_suffix: join with a prefix and suffix.
* join_prefix_suffix_infix: join with a prefix, suffix, and infix.


## Class

* publicize_methods: make all methods public for a block, e.g. to unit test private methods


## CSV

* http_headers: provides web file download headers for text/csv content type and disposition.


## Date

* age_days, age_years
* between: a random date between two specified dates
* to_sql: date as a string formatted as expected for MySQL
* weekday?, weekend?: is date a weekday or on the weekend


## Enumerable

* cartesian_product: return an array of all possible ordered tuples from arrays.
* hash_by: convert the object to a hash by mapping each item to a key=>value pair.
* index_by: convert the object to a hash by mapping each item to a key=>item pair.
* join: forwards to self.to_a.join
* to_h: convert the object to a hash by mapping each item to a key=>value pair.
* to_h_merge: `to_h` and merge any duplicate keys.
* power_set: return an array with all subsets of the enum's elements


### Enumerable map

* map_id: return the id of an Enumerable object; *requires* that the object respond to an 'id' message
* map_to_a, map_to_f, map_to_i, map_to_s, map_to_sym: convert each object to a specific type by calling its respective method to_a, to_i, to_f, to_s, to_sym
* map_with_index: for each item, yield to a block with the item and its incrementing index


### Enumerable nitems

* nitems_while: the number of leading elements for which block is truthy.
* nitems_until: the number of leading elements for which block is falsey.
* nitems_with_index: the number of leading elements for which block with index is truthy.


### Enumerable select

* select_while: return an array of the leading elements for which block is truthy.
* select_until: return an array of the leading elements for which block is falsey.
* select_with_index: return an array of the leading elements for which block with index is truthy.


## File

* File.joindir: wrapper for File.join(File.dirname(...),string,...)


## Fixnum

* even?: is the number even?
* odd?: is the number odd?


## Hash

* size?: return true if hash has any keys
* sort_by_keys: return a new Hash sorted by keys
* each_sort: sort the keys then call each
* each_key!: pass each key to a block; update hash in place with changes.
* each_pair!: pass each key value pair to a block; update hash in place with changes.
* each_value!: pass each value to a block; update hash in place with changes.
* map_key: map each key-value pair's key by calling a a block
* map_pair: map each key-value pair by calling a a block
* map_value: map each key-value pair by calling a a block
* merge_recurse: merge two hashes plus recurse whenever a key is a hash.
* pivot: aggregates subtotals by keys and values, such as a rollup and rolldown
* yield_pair: yield each key-value pair by calling a a block


## Integer

* rbit: reverse bit
* maps: syntactic sugar to yield n times to a block, returning an array of any results
* odd?: is the number odd?


## IO

* readrow: reads a row line as with IO#readline, and return the row split it into fields
* IO.readrows: reads a file and splits text to rows and fields as a 2D array.


## Kernel

* my_method_name: return the name of the current method
* caller_method_name: return the name of the caller method, or the Nth parent up the call stack.


## Math

* ln(x): natural log of x
* logn(x,b): log of x in base b


## NilClass

* blank?: return true (same as Rails)


## Numeric

* if: return 0 if the passed flag is any of: nil, false, 0, [], {} and otherwise return self
* unless: return 0 unless the passed flag is any of: nil, false, 0, [], {} and otherwise return self
* percent: return the number as a percentage with optional rounding to decimal points
* floor_precision: return the number truncated to a given precision, i.e. decimal points


## Object

* in?: return boolean indicating whether the object is a member of the specified array parameter


## Process

Extensions that help debug Ruby programs.

* (class) ps: output of the system 'ps' command, also including aliases, as raw plain text.
* (class) pss: output of the system 'ps' command as a hash with each value set to the right type, e.g., integer, float, etc..


## REXML::Attributes

* hash: flattens the attributes hash set into a more useful ruby hash, e.g. {:height => 100, :width => 400 }


## REXML::Document

* remove_attributes: remove all the attributes from the document's elements


## REXML::Element

* remove_attributes: remove all the attributes from the element


## String

* capitalize_words (alias to titleize/titlecase): ensures the first character of each word is uppercase.
* decrement: decrease the rightmost natural number, with step option.
* increment: increase the rightmost natural number, with step option.
* lorem: return a short random string, good for use in "lorem ipsum" sample text.
* lowcase: translate a string to lowercase, digits and single underscores (e.g. to a method name)
* prev/pred: previous string ("b" => "a", "bbc" => "bbb", "a" => "z", "880" => "879")
* prev!/pred!: updates variable to the previous string in place (astring = "bbc", astring.prev!, puts astring => "bbb")
* (class) prev_char/pred_char: return the previous character, with a changed flag and carry flag; that is, there are three returned values, a string and two booleans.
* split_tab: split the string into an array at each embedded tab ("Last\tFirst\tMiddle" => ["last","first","middle"])
* split_tsv: split the string into an array at each embedded newline and embedded tab ("A\tB\t\C\nD\tE\tF" => [["A","B","C"],["D","E","F"]])
* to_class: the global class reference of a given string
* words: split the string into an array of words


## Symbol

* <=> and include the comparable mixin to compare symbols as strings


## Time

* (class) stamp: current time in UTC as a timestamp string ("YYYYMMDDHHMMSS")
* to_sql: time as a string formatted as expected for MySQL


## YAML

* (class) load_dir: specify directory patterns and pass each YAML file in the matching directories to a block; see [Dir#glob](http://www.ruby-doc.org/core/classes/Dir.html#M002347) for pattern details.
* (class) load_dir_pairs: specify directory patterns and pass each YAML file in the matching directories to a block to process key/value pairs.


CHANGES.md
# Changes

* 2019-04-30 5.0.2 Fix testing of removed method
* 2019-01-27 5.0.1 Fix testing bugs and warnings
* 2018-06-22 5.0.0 Update to Ruby 2.5, testing gems, etc.
* 2017-08-15 4.2.7 Fix deprecation by using assert_nil
* 2017-08-14 4.2.6 Bump minitest expections
* 2017-08-13 4.2.5 Update to Ruby 2.4, spdx licenses, rake test, etc.
* 2015-07-18 4.2.4 Improve quality assurance, tests, and coverage
* 2015-07-07 4.2.3 Update gemspec to use file manifest
* 2015-03-14 4.2.1 Conditionally load `Array#shuffle*`
* 2015-03-14 4.2.0 Add `Integer#rbit`
* 2015-01-13 4.1.0 Fix Enumerable `to_h` and add `to_h_merge`
* 2015-01-13 4.0.1 Fix for Process.ps keys
* 2014-12-17 4.0.0 Add `Pairable` methods; change semantics of `Hash#map_pair`; change `to_h` for Ruby 2.2.0
* 2013-08-19 3.0.1 Update for Code Climate, Travis CI, gem dependencies
* 2013-08-18 3.0.0 Update to Ruby 2
* 2012-09-01 2.1.7 Add Numeric#floor_precision
* 2012-03-14 2.1.6 Update docs, tests
* 2.1.6 Refactor XML#strip* methods to new sixarm_ruby_xml_strip gem
* 2.1.5 Refactor XML#load* methods to new sixarm_ruby_xml_load gem
* 2.1.4 Refactor REXML extensions to new sixarm_ruby_rexml gem
* 2.1.3 Remove dependency on sixarm_ruby_array_slice
* 2.1.2 Refactor Array#slices and #divvy to sixarm_ruby_array_slice gem with better method names
* 2.1.0 Upgrades for Ruby 1.9.3; better README, new CHANGELOG, less YAML, etc.
* 2.0.x Upgrades for Ruby 1.9.2; add methods; lift some methods to their own gems
* 1.8.x 100% rcov coverage; add methods
* 1.7.x Enable gemcutter; add methods; remove sqlite dependency
* 1.6.x Upgrades for Ruby 1.9.1; improve structure; add methods
* 1.5.0 Combined all Ruby extension files into one gem
* 1.0.0 Original


lib/sixarm_ruby_ramp/hash.rb
# -*- coding: utf-8 -*-

require "yaml"
require_relative 'pairable'

# Hash extensions

class Hash
  include Pairable

  # Does the hash contain any items?
  #
  # @return [Boolean] true if size > 0
  #
  def size?
    size>0
  end

  # Sort the hsh items by the keys.
  #
  # @return [Hash] a new hash sorted by the keys
  #
  # @example
  #    h = {"c" => "cherry", "b" => "banana", "a" =>"apple" }
  #    h.sort_keys => {"a" => "apple", "b" => "banana", "c" => "cherry"}
  #
  def sort_by_keys
    Hash[sort]
  end

  # Return a new hash containing the contents of other_hash and the
  # contents of hsh.
  #
  # If no block is specified, the value for entries with duplicate
  # keys will be that of other_hash; if the values are both hashes,
  # then this method recurses.
  #
  # Otherwise the  value for each duplicate key is determined by calling
  # the block with the key, its value in hsh and its value in other_hash.
  #
  # @example:
  #
  #   h1 = {a: 'b', c: {d: 'e'}}
  #   h2 = {a: 'B', c: {d: 'E'}}
  #   h1.merge_recurse(h2)
  #   #=> {a: 'B', c: {d: 'E'}}
  #
  def merge_recurse(other_hash)
    merge(other_hash){|key, oldval, newval|
      if oldval.is_a?(Hash) && newval.is_a?(Hash)
        oldval.merge_recurse(newval)
      else
        newval
      end
    }
  end

  # Hash#pivot aggregates values for a hash of hashes,
  # for example to calculate subtotals and groups.
  #
  # Suppose you have data arranged by companies, roles, and headcounts.
  #
  #     data = {
  #     "Apple"     => {"Accountants" => 11, "Designers" => 22, "Developers" => 33},
  #     "Goggle"    => {"Accountants" => 44, "Designers" => 55, "Developers" => 66},
  #     "Microsoft" => {"Accountants" => 77, "Designers" => 88, "Developers" => 99},
  #   }
  #
  # To calculate each company's total headcount, you pivot up, then sum:
  #
  #   data.pivot(:up,&:sum)
  #   => {
  #    "Apple"=>66,
  #    "Goggle"=>165,
  #    "Microsoft"=>264
  #   }
  #
  # To calculate each role's total headcount, you pivot down, then sum:
  #
  #   data.pivot(:down,&:sum)
  #   => {
  #    "Accountants"=>132,
  #    "Designers"=>165,
  #    "Developers"=>198
  #   }
  #
  #  Generic example:
  #   h={
  #    "a"=>{"x"=>1,"y"=>2,"z"=>3},
  #    "b"=>{"x"=>4,"y"=>5,"z"=>6},
  #    "c"=>{"x"=>7,"y"=>8,"z"=>9},
  #   }
  #   h.pivot(:keys) => {"a"=>[1,2,3],"b"=>[4,5,6],"c"=>[7,8,9]}
  #   h.pivot(:vals) => {"x"=>[1,4,7],"y"=>[2,5,8],"z"=>[3,6,9]}
  #
  # = Calculating subtotals
  #
  # The pivot method is especially useful for calculating subtotals.
  #
  # @example
  #   r = h.pivot(:keys)
  #   r['a'].sum => 6
  #   r['b'].sum => 15
  #   r['c'].sum => 24
  #
  # @example
  #   r=h.pivot(:vals)
  #   r['x'].sum => 12
  #   r['y'].sum => 15
  #   r['z'].sum => 18
  #
  # = Block customization
  #
  # You can provide a block that will be called for the pivot items.
  #
  # @example
  #   h.pivot(:keys){|items| items.max } => {"a"=>3,"b"=>6,"c"=>9}
  #   h.pivot(:keys){|items| items.join("/") } => {"a"=>"1/2/3","b"=>"4/5/6","c"=>"7/8/9"}
  #   h.pivot(:keys){|items| items.inject{|sum,x| sum+=x } } => {"a"=>6,"b"=>15,"c"=>24}
  #
  # @example
  #   h.pivot(:vals){|items| items.max } => {"a"=>7,"b"=>8,"c"=>9}
  #   h.pivot(:vals){|items| items.join("-") } => {"a"=>"1-4-7","b"=>"2-5-8","c"=>"3-6-9"}
  #   h.pivot(:vals){|items| items.inject{|sum,x| sum+=x } } => {"a"=>12,"b"=>15,"c"=>18}

  def pivot(direction='keys',&block)
    a=self.class.new
    direction=direction.to_s
    up=pivot_direction_up?(direction)
    each_pair{|k1,v1|
      v1.each_pair{|k2,v2|
        k = up ? k1 : k2
        a[k]=[] if (a[k]==nil or a[k]=={})
        a[k]<<(v2)
      }
    }
    if block
      a.each_pair{|key,val| a[key]=block.call(val)}
    end
    a
  end

  protected

  def pivot_direction_up?(direction_name)
    case direction_name.to_s
    when 'key','keys','up','left','out' then return true
    when 'val','vals','down','right','in' then return false
    else raise ArgumentError.new('Pivot direction must be either: key/keys/up/left/out or val/vals/down/right/in')
    end
  end

end
